File: day25/day25.py
MOD=20201227
SUBJ=7

A=6930903
B=19716708

# Example:
assert pow(17807724, 8, MOD)==14897079
assert pow(5764801, 11, MOD)==14897079

a,b=1,1
secret = 0
while True:
  if secret%1000000 == 0: print('...',secret)
  if a==A:
    print('a',a,A,secret)
    print('1--->',pow(B,secret,MOD))
    break
  if b==B:
    print('b',b,B,secret)
    print('1--->',pow(A,secret,MOD))
    break
  a = (a*SUBJ)%MOD
  b = (b*SUBJ)%MOD
  secret += 1

File: day25/day25.c
#include <stdio.h>

int MOD=20201227;
int SUBJ=7;

int A=6930903;
int B=19716708;

long powm(long base, unsigned long exp, long modulus) {
  base %= modulus;
  long result = 1;
  while (exp > 0) {
    if (exp & 1) result = (result * base) % modulus;
    base = (base * base) % modulus;
    exp >>= 1;
  }
  return result;
}

int main() {
  int a=1,b=1;
  int secret=0;
  while(1) {
    if (secret%1000000 == 0) printf("...%d\n",secret);
    if (a==A) {
      printf("a %d,%d: %d\n",a,A,secret);
      printf("1--->%d\n",powm(B,secret,MOD));
      break;
    }
    if (b==B) {
      printf("b %d,%d: %d\n",b,B,secret);
      printf("1--->%d\n",powm(A,secret,MOD));
      break;
    }
    a = (a*SUBJ)%MOD;
    b = (b*SUBJ)%MOD;
    ++secret;
  }
}
